security/pkg/nodeagent/caclient/providers/spire/client.go

```diff
@@ -0,0 +1,358 @@
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+	"sync"
+
+	spiffeTls "github.com/spiffe/go-spiffe/tls"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	clientInterface "istio.io/istio/security/pkg/nodeagent/caclient/interface"
+	spireIntegration "istio.io/istio/security/proto/providers/spire"
+	"istio.io/pkg/log"
+)
+
+const (
+	serverID         = "spiffe://%s/spire/server"
+	spiffeIDTemplate = "spiffe://%s"
+)
+
+type spireClient struct {
+	serverAddr  string
+	trustDomain string
+
+	client       spireIntegration.NodeClient
+	clientConn   *grpc.ClientConn
+	latestBundle string
+	mutex        *sync.Mutex
+}
+
+// NewSpireClient create a CA client for Spire.
+func NewSpireClient(serverAddr string, tlsRootCert []byte, trustDomain string) (clientInterface.Client, error) {
+	c := &spireClient{
+		serverAddr:   serverAddr,
+		trustDomain:  trustDomain,
+		latestBundle: string(tlsRootCert),
+		mutex:        &sync.Mutex{},
+	}
+
+	err := c.createNodeClient(tlsRootCert)
+	if err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+type IstioAttestableData struct {
+	Token string `json:"token"`
+}
+
+// CSRSign calls Spire server to sign a CSR.
+func (c *spireClient) CSRSign(ctx context.Context, csrPEM []byte, token string,
+	certValidTTLInSec int64) ([]string, error) {
+
+	attestableData := &IstioAttestableData{
+		Token: token,
+	}
+
+	data, err := json.Marshal(attestableData)
+	if err != nil {
+		log.Errorf("Failed to parse attestable SPIRE data %v", err)
+		return nil, errors.New("failed to parse attestable SPIRE data")
+	}
+	attestationData := &spireIntegration.AttestationData{
+		Data: data,
+		Type: "istio",
+	}
+
+	certRequest, err := csrFromPEM(csrPEM)
+	if err != nil {
+		log.Errorf("Could not decode csr to be signed by SPIRE: %v", err)
+		return nil, errors.New("could not decode csr to be signed by SPIRE")
+	}
+
+	if len(certRequest.URIs) != 1 {
+		return nil, errors.New("signing certificate must contain exactly one SPIFFE ID")
+	}
+
+	certID := certRequest.URIs[0]
+	if err := c.validateSpiffeID(certID); err != nil {
+		log.Errorf("Spiffe id from CSR is not valid: %v", err)
+		return nil, errors.New("spiffe id from CSR is not valid")
+	}
+
+	attestReq := &spireIntegration.AttestRequest{
+		Csr:             certRequest.Raw,
+		AttestationData: attestationData,
+	}
+
+	// cancel context to be sure stream is closed
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// create an attestation stream
+	attestStream, err := c.client.Attest(ctx)
+	if err != nil {
+		log.Errorf("Failed opening stream to attestating SPIRE server: %v", err)
+		return nil, errors.New("failed opening stream to attestating SPIRE server")
+	}
+
+	if err := attestStream.Send(attestReq); err != nil {
+		log.Errorf("Not able to send attestation request to SPIRE server: %v", err)
+		return nil, fmt.Errorf("not able to send attestation request to SPIRE server")
+	}
+
+	attestResp, err := attestStream.Recv()
```

Comment:
Just curious, is your CA push model ? it looks like it's stream based which means client will keep getting update whenever your CA push new data

Reply:
The Attest() RPC is implemented as a bidirectional stream because some attestation methods supported by SPIRE require challenge/response. Once attestation is complete, Attest() returns the SVID and the stream is closed. It is not used to continue streaming updates.